src/Stratis.Bitcoin/Consensus/PHConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,306 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Connection;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.P2P.Protocol;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    public class PHConsensusManagerBehavior : ConsensusManagerBehavior
+    {
+        private readonly ConcurrentChain chain;
+        private readonly IInitialBlockDownloadState initialBlockDownloadState;
+        private readonly IConsensusManager consensusManager;
+        private readonly IPeerBanning peerBanning;
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Protects write access to the <see cref="BestSentHeader"/>.</summary>
+        private readonly object bestSentHeaderLock;
+
+        /// <summary>Gets the best header sent using <see cref="ProvenHeadersPayload"/>.</summary>
+        /// <remarks>Write access should be protected by <see cref="bestSentHeaderLock"/>.</remarks>
+        public ProvenBlockHeader BestSentHeader { get; private set; }
+
+        /// <summary>Maximum number of headers in <see cref="HeadersPayload"/> according to Bitcoin protocol.</summary>
+        /// <seealso cref="https://en.bitcoin.it/wiki/Protocol_documentation#getheaders"/>
+        private const int MaxItemsPerHeadersMessage = 2000;
+
+        public PHConsensusManagerBehavior(ConcurrentChain chain, IInitialBlockDownloadState initialBlockDownloadState, IConsensusManager consensusManager, IPeerBanning peerBanning, ILoggerFactory loggerFactory) : base(chain, initialBlockDownloadState, consensusManager, peerBanning, loggerFactory)
+        {
+            this.chain = chain;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.peerBanning = peerBanning;
+            this.loggerFactory = loggerFactory;
+            this.bestSentHeaderLock = new object();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        /// <summary>
+        /// Processes and incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
```

Comment:
Override base.OnMessageReceivedAsync instead

Reply:
You mean override the base method not `Override base.OnMessageReceivedAsync`
Yeah agree this is preffered